src/web/statics.rs

```diff
@@ -23,29 +23,37 @@ pub(crate) fn static_handler(req: &mut Request) -> IronResult<Response> {
         "vendored.css" => serve_resource(VENDORED_CSS, ContentType("text/css".parse().unwrap()))?,
         "style.css" => serve_resource(STYLE_CSS, ContentType("text/css".parse().unwrap()))?,
         "rustdoc.css" => serve_resource(RUSTDOC_CSS, ContentType("text/css".parse().unwrap()))?,
-        file => serve_file(req, file)?,
+        file => serve_file(file)?,
     })
 }
 
-fn serve_file(req: &Request, file: &str) -> IronResult<Response> {
+fn serve_file(file: &str) -> IronResult<Response> {
     // Find the first path that actually exists
     let path = STATIC_SEARCH_PATHS
         .iter()
         .filter_map(|root| {
             let path = Path::new(root).join(file);
+            if !path.exists() {
+                return None;
+            }
 
             // Prevent accessing static files outside the root. This could happen if the path
             // contains `/` or `..`. The check doesn't outright prevent those strings to be present
             // to allow accessing files in subdirectories.
-            if path.starts_with(root) {
-                Some(path)
+            let canonical_path = std::fs::canonicalize(path).ok()?;
+            let canonical_root = std::fs::canonicalize(root).ok()?;
+            if canonical_path.starts_with(canonical_root) {
+                Some(canonical_path)
             } else {
                 None
             }
         })
-        .find(|p| p.exists())
+        .next()
         .ok_or(Nope::ResourceNotFound)?;
-    let contents = ctry!(req, fs::read(&path));
+    let contents = fs::read(&path).map_err(|e| {
+        log::error!("failed to read static file {}: {}", path.display(), e);
+        Nope::InternalServerError
```

Comment:
This doesn't look correct; it will give internal server errors when you try to request a file that doesn't exist. Can you either

a) add back the `exists()` call, or
b) check if `e.kind() == ErrorKind::NotFound`

and either way add a test?

Reply:
The `exists()` call is already there, on line 35.